### ISSUE TITLE
Improve Python environment activation

### ISSUE THREAD

**Issue by mattseddon** [iterative/vscode-dvc#1791]
Currently for Python virtual environments we use `ms-python.python`'s API to grab the currently selected Python interpreter and then we run:

`$python_bin -c "import sys; print(sys.executable)"`

This gives us the absolute path of the interpreter. We then call `dvc` with `path/to/interpreter/python -m dvc with some args`. We also prepend the directory that the python interpreter is in to the system $PATH variable before running the command. This works for the basic case of DVC being installed within the virtual environment and is [the suggested approach](https://github.com/microsoft/vscode-python/discussions/18978#discussioncomment-2618085) from an ongoing github discussion. 

Our current approach also works for global installs of DVC as long as the required dependencies are installed within the virtual environment. E.g for the demo project a global install of DVC does not work unless `dvclive` and `ruamel.yaml` are added to the virtual environment. I don't think there is a way around this for the brew installation. I looked into what can be done for a global install via `pip` but I can recreate the same behaviour via the terminal after activating the virtual environment.

I've been looking into how other extensions do this again and have found the following issues/discussions:

- https://github.com/microsoft/vscode-python/issues/11039
- https://github.com/microsoft/vscode-python/discussions/18978
- https://github.com/guildai/guildai-vscode/issues/1
- https://github.com/brettcannon/python-launcher/discussions/168

This seems to be an area that the `vscode-python` team is actively working on. We should keep an eye on those discussion tickets and see where they go. If they could expose the activation command in their API that would be ideal. Maybe we can get the correct PATH details from [EnvironmentVariableCollection](https://github.com/microsoft/vscode/blob/17c4bd39f0fc49886004a2685b8350ab541a1e4d/src/vscode-dts/vscode.d.ts#L10769) in the extension context.

**Comment by erykml** [comment]
Hi! I was directed here by @shcheklein. I am running into a problem with the DVC extension and virtual envs. 

I have a codebase, and set up a `dvc.yaml` pipeline, which works. I tried `dvc repro` command (with the activated conda env) and everything succeeded. Then, I wanted to run an experiment and used `dvc exp run`, which also succeeded. Unfortunately, that is not the case when using the VS Code extension. 

Then, I am getting an error `ModuleNotFoundError: No module named 'src'`, which is weird as the codebase did not change between `repro` and `exp run`. 

A potential culprit might be the pythonpath setting. In VS Code, I have an `.env` file with `PYTHONPATH=src` and in `settings.json` I have the following: `"python.envFile": "${workspaceFolder}/.env"`. I added is as a suggestion from one of the users of DVC discord chat, but unfortunately it did not solve the problem.

To sum up, I am using macOS and working with a conda environment.

Any ideas about what might be the source of the issue and how to solve it?

**Comment by erykml** [comment]
Another related  issue that pops up while playing around with the extension is that once I try using the experiments tab (it does not work as mentioned above) to run an experiment, I can't run them anymore from the terminal using `dvc exp run`. I need to run `export PYTHONPATH=$PWD` in order to be able to run the experiments from the terminal again. This hints that the extension is unsuccessfully trying to set the path.

**Comment by shcheklein** [comment]
@erykml 

> I can't run them anymore from the terminal using dvc exp run

Does it depend on whether extension is active or not? 

Could you also please share the project layout and how do you run your script usually? I'm curios why do we need to do `PYTHONPATH` at all tbh. There are ways to do something like `python train.py` w/o providing PYTHONPATH.

**Comment by erykml** [comment]
@shcheklein 

After I tried running the `dvc exp run` command in the integrated terminal of VS Code after the failed run from the extension, then I am getting the error. I have killed the extensions terminal window by then. Does this answer your question?

The project's structure is as follows:

```
# the structure of the project

📦vscode_exp_tracking_with_dvc
 ┣ 📂.dvc
 ┣ 📂data
 ┃ ┣ 📂processed
 ┃ ┣ 📂raw
 ┃ ┣ 📜.gitignore
 ┃ ┗ 📜raw.dvc
 ┣ 📂models
 ┣ 📂src
 ┃ ┣ 📂stages
 ┃ ┃ ┣ 📜data_split.py
 ┃ ┃ ┣ 📜eval.py
 ┃ ┃ ┗ 📜train.py
 ┃ ┗ 📂utils
 ┃ ┃ ┣ 📜__init__.py
 ┃ ┃ ┗ 📜params.py
 ┣ 📜.dvcignore
 ┣ 📜.gitignore
 ┣ 📜README.md
 ┣ 📜dvc.lock
 ┣ 📜dvc.yaml
 ┣ 📜metrics.json
 ┣ 📜params.yaml
 ┗ 📜requirements.txt
```

And normally I run a script as `python src/stages/train.py`.

**Comment by shcheklein** [comment]
Got it. 

> This hints that the extension is unsuccessfully trying to set the path.

This is interesting. We'll check.

> And normally I run a script as python src/stages/train.py.

thanks, it makes sense. Why do you need to use `PYTHONPATH`? I think you can import modules / scripts properly in the `train.py` so that `PYTHONPATH` is not required?

**Comment by erykml** [comment]
Currently, I import the modules/scripts as follows:
```
import pandas as pd
from lightgbm import LGBMClassifier
from sklearn.ensemble import RandomForestClassifier

from src.utils.params import load_params
```

**Comment by shcheklein** [comment]
@SoyGema could you please also add a summary of the road block that you hit with the env vars / PYTHONPATH?

**Comment by SoyGema** [comment]
**CONTEXT** : Reproducing the experiments for [workshop](https://github.com/iterative/VSCode-DVC-Workshop) , with `dvc.yaml` and `params.yaml` . 

When executing `dvc exp run train `, the CLI is throwing the error:

```
Running stage 'train':                                                
> python  ExperimentsDVC/src/models/train_model.py
Traceback (most recent call last):
  File "/path/ExperimentsDVC/src/models/train_model.py", line 2, in <module>
    from src.data.load_dataset import read_data
ModuleNotFoundError: No module named 'src'
ERROR: failed to reproduce 'train': failed to run: python  ExperimentsDVC/src/models/train_model.py, exited with 1
```
That leads to set PYTHONPATH in CLI  with  `export PYTHONPATH = 'path/ExperimentsDVC'`
With this, dvc experiments is able to take the absolute path and train the experiment , showing metrics in table.

Then, pressing left button and executing `Modify and Run`  UI feature , and changing some params , the following message is shown.

```
>>>> DVC Terminal >>>>

Running: dvc exp run -S params.yaml:train.n_estimators=150 -S params.yaml:train.max_depth=25 -S params.yaml:train.min_samples_split=10 -S params.yaml:load.data_path_train=ExperimentsDVC/data/raw/jan_train.csv -S params.yaml:load.data_path_test=ExperimentsDVC/data/raw/jan_test.csv -S params.yaml:featureselection.data=ExperimentsDVC/data/raw/jan_train.csv -S params.yaml:featureselection.features=["id","sat_id","x_sim","y_sim","z_sim","Vx_sim","Vy_sim","Vz_sim","Day","Hour","Minute"] -S params.yaml:featureselection.labels=["x","y","z","Vx","Vy","Vz"] -S params.yaml:splitdata.size=0.33 -S params.yaml:splitdata.state=42 -S params.yaml:standardscaler.data=null -S params.yaml:train.min_impurity_decrease=1

Running stage 'train':
> python  ExperimentsDVC/src/models/train_model.py
Traceback (most recent call last):
  File "/path/VSCode-DVC-Workshop/ExperimentsDVC/src/models/train_model.py", line 2, in <module>
    from src.data.load_dataset import read_data
ModuleNotFoundError: No module named 'src'
ERROR: failed to reproduce 'train': failed to run: python  ExperimentsDVC/src/models/train_model.py, exited with 1
```
Which seems similar to the previous error , which might lead to "assume" or "hypothesize" that the Workspace UI path config doesn´t relate to defined PYTHONPATH in CLI

**CONFIG** 
Virtual env with [requirements](https://github.com/iterative/VSCode-DVC-Workshop/blob/main/requirements.txt) . Setup Workspace with **DVC:Setup The Workspace**  > Auto . Bottom Left DVC Circle UI complete and bottom right Python seems correct 3.9.13('env': venv)

**Comment by mattseddon** [comment]
> Looks like PYTHONPATH is now exposed by the Python Extension: https://github.com/microsoft/vscode-python/blob/main/src/client/api.ts#L80

This is however marked as deprecated so we need to wait for it to get moved into stable before we go ahead and use it: https://github.com/microsoft/vscode-python/blob/main/src/client/api.ts#L30
